### PR TITLE
[Accuracy diff No.25] Fix accuracy diff for paddle.linalg.eigh API

### PR DESCRIPTION
**Pull request by NKNaN**
问题的原因是实对称矩阵或hermitian matrix的特征向量是不唯一的，可能因为底层实现的不同导致了 paddle 和 torch 的输出不同，所以应该对 paddle.linalg.eigh 的特征向量输出做单独的精度判断：一个简单的判断办法是将 paddle 和 torch 对应输出的特征向量相除，得到的应该是  e^{i \theta} *  vec[1, 1, ..., 1]^T  ，所以判断相除所得向量的所有元素(大致)相等，以及相除所得向量每个元素的模长为1即可。

另外由于 eigh_grad 的计算依赖于前向计算得到的特征向量，因此反向计算的结果也会有差异，所以建议不测 eigh 的反向算子
image

回测结果：
GPU:
image
CPU:
image

### CODE REVIEW

**Comment by cangtianhuang** [comment]
@NKNaN 同学好😊！这应该是属于 paddle 与 torch 在设计上不一致，我们 apitest 要做的就是尽量将 torch 的行为与 paddle 对齐～

如果两者的差异在理论上是可以接受的（均是正确的特征向量），我们就需要修改 torch 的输出，不改 paddle 的输出，具体修改应该属于 paddle2torch 模块，位于：tester/paddle_to_torch

需要的修改有：
- mapping.json：为 paddle.linalg.eigh 注册 Rule
- rules.py 编写 rule 类，在后处理 post 中修正 torch 的输出，使其与 paddle 对齐

这样即可正常进行测试，不必修改 tester/accuracy.py～～

**Comment by NKNaN** [comment]
嗯嗯了解，但是问题是输出的特征向量是拼接在一起的一个矩阵，可能有的特征向量是一样的有的是不一样的，而且好像也没有一个特定的规律，比如得到的特征向量矩阵 paddle 是 [v1^T, v2^T, v3^T, v4^T]，但 torch 可能是 [-v1^T, v2^T, -v3^T, v4^T] 或者 [v1^T, -v2^T, v3^T, -v4^T] 或者 [e^{i \theta_1} * v1^T,  e^{i \theta_2} * v2^T, e^{i \theta_3} * v3^T, v4^T] (theta 可以是任意角度)都有可能，这样的话好像也找不到一个统一的规则去修改 torch 的输出？

**Comment by NKNaN** [comment]
> 哦～我明白你的意思了，那paddle与torch输出的特征向量顺序是否一定相同呢？是否可能有 paddle 是 [v1^T, v2^T, v3^T, v4^T]， torch 是 [-v2^T, v1^T, -v4^T, v3^T]？ 

特征向量顺序是一致的，和所得特征值的相对顺序一样。

> 此外，在此处直接torch_assert_accuracy处理，不利于走完全流程，同学可以考虑做做如下处理，让这个api显得不那么特立独行：
> 
> 1. `paddle_output, torch_output` 设置为 `[coef_vector, abs_coef], [coef_vector_approx, one]`
> 2. 将 `paddle.linalg.eigh` 添加进 `tester/base.py` 的 `forward_only_apis` 列表中

修改了一下，回测也可以通过。

File: tester/accuracy.py [hunk 1]
```diff
@@ -233,6 +233,28 @@ def test(self):
             # torch's from_dlpack now don't support negative strides
             paddle_output = paddle_output.contiguous()
 
+        paddle_output_cache = []
```
Comment (cangtianhuang) [inline tester/accuracy.py:236]:
```suggestion
```

File: tester/accuracy.py [hunk 2]
```diff
@@ -233,6 +233,28 @@ def test(self):
...
+            torch_eigvectors = torch_output.pop(1).transpose(-1, -2).reshape((-1, eigvector_len))
+            for i in range(paddle_eigvectors.shape[0]):
+                coef_vector = paddle.to_tensor(paddle_eigvectors[i].numpy()/torch_eigvectors[i].numpy(), dtype=paddle_eigvectors[i].dtype)
+                coef_vector = coef_vector.round(2)
+                coef_0 = paddle_eigvectors[i].numpy()[0]/torch_eigvectors[i].numpy()[0]
+                coef_vector_approx = torch.tensor([coef_0] * eigvector_len)
+                abs_coef = coef_vector.abs().astype("float64")[0]
+                one = torch.tensor(1.0, dtype=torch.float64)
+                paddle_output.append([coef_vector, abs_coef])
+                torch_output.append([coef_vector_approx, one])
```
Comment (cangtianhuang) [inline tester/accuracy.py:255]: 我的想法是：
```
paddle_output = [coef_vector, abs_coef]
torch_output = [coef_vector_approx, one]
```
这样可以同时比较绝对误差和相对误差，不需要 paddle_output_cache～

File: tester/accuracy.py [hunk 3]
```diff
@@ -233,6 +233,28 @@ def test(self):
...
             paddle_output = paddle_output.contiguous()
 
+        paddle_output_cache = []
+        if self.api_config.api_name == "paddle.linalg.eigh":
+            # The output of eigen vectors are not unique, because multiplying an eigen vector by -1 in the real case 
+            # or by e^(i*\theta) in the complex case produces another set of valid eigen vectors of the matrix.
+            # So we test whether the elements of each coef_vector (i.e. paddle_output / torch_output for each eigen vector) 
+            # are all the same and whether the |coef| == 1 for simplicity.
+            paddle_output, torch_output = list(paddle_output), list(torch_output)
+            paddle_output_cache = [i.clone() for i in paddle_output]
```
Comment (cangtianhuang) [inline tester/accuracy.py:243]:
```suggestion
```

File: tester/accuracy.py [hunk 4]
```diff
@@ -332,6 +354,9 @@ def test(self):
                             write_to_log("accuracy_error", self.api_config.config)
                             return
 
+        if self.api_config.api_name == "paddle.linalg.eigh":
+            paddle_output = paddle_output_cache
+
```
Comment (cangtianhuang) [inline tester/accuracy.py:359]:
```suggestion
```

**Review by cangtianhuang — CHANGES_REQUESTED** [review]

File: tester/accuracy.py (hunk 2 above)
Comment (NKNaN) [reply tester/accuracy.py:253]: 嗯嗯，忘记删了

**Review by cangtianhuang — APPROVED** [review]
LGTM

**Comment by NKNaN** [comment]
@ luotao1
No. 101 也是 paddle.linalg.eigh, 之前忘记加在标题里了~

**Comment by cangtianhuang** [comment]
> @ luotao1 No. 101 也是 paddle.linalg.eigh, 之前忘记加在标题里了~

直接修改标题就行～
